controller/volume_rebuilding_controller.go

```diff
@@ -96,6 +120,82 @@ func (vbc *VolumeRebuildingController) enqueueVolumeAfter(obj interface{}, durat
 	vbc.queue.AddAfter(key, duration)
 }
 
+func isSettingOfflineReplicaRebuilding(obj interface{}) bool {
+	setting, ok := obj.(*longhorn.Setting)
+	if !ok {
+		return false
+	}
+	return setting.Name == string(types.SettingNameOfflineReplicaRebuilding)
+}
+
+func (vbc *VolumeRebuildingController) enqueueSetting(obj interface{}) {
+	_, ok := obj.(*longhorn.Setting)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("DeletedFinalStateUnknown contained non Setting object: %#v", obj))
+		return
+	}
+
+	vs, err := vbc.ds.ListVolumesRO()
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to list volumes: %v", err))
+		return
+	}
+
+	for _, v := range vs {
+		if v.Spec.OfflineRebuilding != longhorn.VolumeOfflineRebuildingIgnored {
+			continue
+		}
+		vbc.enqueueVolume(v)
+	}
+}
+
+func isReplicaOwnerChanged(obj interface{}) bool {
+	r, ok := obj.(*longhorn.Replica)
+	if !ok {
+		return false
+	}
+	if !r.DeletionTimestamp.IsZero() {
+		return true
+	}
+
+	return r.Spec.NodeID != r.Status.OwnerID
+}
+
+func (vbc *VolumeRebuildingController) enqueueVolumeForReplica(obj interface{}) {
+	r, ok := obj.(*longhorn.Replica)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("DeletedFinalStateUnknown contained non SetNodeting object: %#v", obj))
+		return
+	}
+
+	v, err := vbc.ds.GetVolumeRO(r.Spec.VolumeName)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to get volume: %v", err))
+		return
+	}
+	if v.Status.State != longhorn.VolumeStateDetached {
+		return
+	}
+
+	n, err := vbc.ds.GetNodeRO(r.Spec.NodeID)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			vbc.enqueueVolume(v)
+			return
+		}
+		utilruntime.HandleError(fmt.Errorf("failed to get node: %v", err))
+		return
+	}
+	if n.Status.Conditions != nil {
+		nodeReadyCondition := types.GetCondition(n.Status.Conditions, longhorn.NodeConditionTypeReady)
+		if nodeReadyCondition.Status == longhorn.ConditionStatusTrue {
+			return
+		}
+	}
```

Comment:
Do we need to check node resource?
Can we check whether each replicas' spec.NodeID is equal to stataus.OwnerID?
If No, the volume should be enqueued.

Reply:
Removed. Thanks.